src/annotate.rs

```diff
@@ -30,17 +35,76 @@ fn protocol_detector(wasm_path: PathBuf) -> Result<ProtocolVersion> {
         .map_err(|e| anyhow!("Cannot compute ProtocolVersion used by the policy: {:?}", e))
 }
 
+// Looks at the Wasm module pointed by `wasm_path` and return whether it was generaed by a Rego
+// policy
+//
+// The code looks at the export symbols offered by the Wasm module.
+// Having at least one symbol that starts with the `opa_` prefix leads
+// the policy to be considered a Rego-based one.
+fn rego_policy_detector(wasm_path: PathBuf) -> Result<bool> {
+    let data: Vec<u8> = std::fs::read(wasm_path)?;
+    for payload in wasmparser::Parser::new(0).parse_all(&data) {
+        if let wasmparser::Payload::ExportSection(s) = payload? {
+            for export in s {
+                if export?.field.starts_with("opa_") {
+                    return Ok(true);
+                }
+            }
+        }
+    }
+
+    Ok(false)
+}
+
 fn prepare_metadata(
     wasm_path: PathBuf,
     metadata_path: PathBuf,
-    detect_protocol_func: impl Fn(PathBuf) -> Result<ProtocolVersion>,
+    kubewarden_protocol_detector_func: impl Fn(PathBuf) -> Result<ProtocolVersion>,
+    rego_detector_func: impl Fn(PathBuf) -> Result<bool>,
```

Comment:
I think I prefer to shuffle this data structures a bit, such as:

```rust
enum Backend {
  Unknown,
  OPA,
  OPAGatekeeper,
  KubewardenWapc(ProtocolVersion),
}
```

And we can have a function that performs the heuristics and return a `Result<Backend>`. By decomposing the result, we get the relevant information in each case -- right now, the protocol version in the case of the kubewarden protocol enum value.

The mocking strategy is similar to this one, in that we can pass this heuristics function as an `impl Fn` argument we can mock.

Reply:
I like the proposal, I'll change the code to reflect that. Thanks!